File: 136_palindrome-partitioning/palindrome-partitioning.cpp
/*
@Copyright:LintCode
@Author:   heiyanbin
@Problem:  http://www.lintcode.com/problem/palindrome-partitioning
@Language: C++
@Datetime: 14-12-08 04:03
*/

class Solution {
public:
    /**
     * @param s: A string
     * @return: A list of lists of string
     */
    vector<vector<string>> partition(string s) {
        // write your code here
        vector<string> path;
        vector<vector<string>> res;
        dfs(s,0,path,res);
        return res;
    }
    void dfs(string &s, int start, vector<string> &path, vector<vector<string>> &res)
    {
        if(start==s.size())
            res.push_back(path);
        for(int i=start;i<s.size();++i)
            if(isPal(s,start,i))
            {
                path.push_back(s.substr(start,i-start+1));
                dfs(s,i+1,path,res);
                path.pop_back();
            }
    }
    bool isPal(string &s,int start, int end)
    {
        for(int L=start, R=end;L<R;L++,R--)
            if(s[L]!=s[R])
                return false;
        return true;
    }
};

File: 136_palindrome-partitioning/palindrome-partitioning.java
/*
@Copyright:LintCode
@Author:   heiyanbin
@Problem:  http://www.lintcode.com/problem/palindrome-partitioning
@Language: Java
@Datetime: 15-10-29 07:36
*/

public class Solution {
    /**
     * @param s: A string
     * @return: A list of lists of string
     */
    public List<List<String>> partition(final String s) {
		final List<List<String>> ans = new ArrayList<List<String>>();
		if (s == null || s.length() == 0) return ans;
		final int n = s.length();
		final boolean[][] isPal = new boolean[n][n];
		for (int i = n - 1; i >= 0; --i) {
			for (int j = i; j < n; ++j) {
				isPal[i][j] = s.charAt(i) == s.charAt(j) && (j - i < 2 || isPal[i + 1][j - 1]);
			}
		}
		class DFS {
			void dfs(int i, ArrayList<String> path) {
				if (i == n) {
					ans.add(new ArrayList<String>(path));
				} else {
					for (int j = i; j < n; ++j) {
						if (isPal[i][j]) {
							path.add(s.substring(i, j + 1));
							dfs(j + 1, path);
							path.remove(path.size() - 1);
						}
					}
				}			
			}
		}
		new DFS().dfs(0, new ArrayList<String>());
		return ans;
	}
}
